web/docs/typescript.md

```diff
@@ -0,0 +1,271 @@
+
+Entity types are also available on the client under the same import:
+```tsx title="Main.js"
+import { Task } from "@wasp/entities"
+
+const task: Task = {
+  id: 123,
+  description: "Some random task",
+  isDone: false,
+}
+
+alert(JSON.stringify(task))
+```
+The mentioned type safety mechanisms also apply here: Changing the task entity in our `.wasp` file changes the imported type, which might throw a type error and warn us that our task definition is outdated.
+
+
+## Backend type support for Operations
+Wasp automatically generates the appropriate types for all operations (i.e., Actions and Queries) you define inside your `.wasp` file. Assuming your `.wasp` file contains the following definition:
+```css
+query GetTaskInfo {
+  fn: import { getTaskInfo } from "@server/queries.js",
+  entities: [Task]
+}
+```
+Wasp will generate a type called `GetTaskInfo`, which you can use to type the Query's implementation. By assigning the `GetTaskInfo` type to your function, you get the type information for its context. In this case, TypeScript will know that the `context.entities` object must include the `Task` entity. If the Query had auth enabled, it would also know that `context` includes user information.
+
+`GetTaskInfo` can is a generic type that takes two (optional) type arguments:
+1. `Input` - The argument (i.e., payload) received by the query function.
+2. `Output` - The query function's return type.
+
+Suppose you don't care about typing the Query's inputs and outputs. In that case, you can omit both type arguments, and TypeScript will infer the most general types (i.e., `never` for the input, `unknown` for the output.).
+
+```typescript
+import HttpError from "@wasp/core/HttpError.js"
+import { Task } from "@wasp/entities"
+// highlight-next-line
+import { GetTaskInfo } from "@wasp/queries/types"
+
+function getInfoMessage(task: Pick<Task, "isDone" | "description">): string {
+  const isDoneText = task.isDone ? "is done" : "is not done"
+  return `Task '${task.description}' is ${isDoneText}.`
+}
+
+// Use the type parameters to specify the Query's argument and return types.
+// highlight-next-line
+export const getTaskInfo: GetTaskInfo<Pick<Task, "id">, string> = async ({ id }, context) => {
+  // Thanks to the definition in your .wasp file, the compiler knows the type of
+  // `context` (and that it contains the `Task` entity).
+  const Task = context.entities.Task
+
+  // Thanks to the first type argument in `GetTaskInfo`, the compiler knows `args`
+  // is of type `Pick<Task, "id">`.
+  const task = await Task.findUnique({ where: { id } })
+  if (!task) {
+    throw new HttpError(404)
+  }
+
+  // Thanks to the second type argument in `GetTaskInfo`, the compiler knows the
+  // function must return a value of type `string`.
+  return getInfoMessage(task)
+}
+```
+Everything described above applies to Actions.
+
+## Frontend type support for Queries and Actions
+### Type support for the `useQuery` hook
```

Comment:
I got a little confused with the section. There is a lot of moving parts:
- request payload
- response payload
- error type
- component props

I would maybe introduce these concepts with two examples e.g. one for the request and response, and then add error and component props.

Reply:
I dropped the props entirely and separated the rest as suggested.